### PR TITLE
Tidy up various thread interactions

### PR DESCRIPTION
**Pull request by garypen**
Introduce the DropWatch utility struct which can be used to watch a closure and either:
 - execute the closure and then wait for the DropWatch struct to be dropped
 - or wait for the struct to be dropped and then execute the closure

It's primarily useful when dealing with Drops that are triggered from an async context, but could be useful wherever we have a thread that we'd like to have a managed lifetime associated with a Drop.

fixes: #4121

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by BrynCooke** [comment]
I think I'm going to need some more explanation as to why this is going to work.
It'll park one of the tokio threads while doing its thing, but why is this OK?

File: apollo-router/src/drop_watch.rs
Comment (BrynCooke) [inline apollo-router/src/drop_watch.rs:22]: How about using a tokio mutex? Create an owned lock guard and pass it into the spawned thread.
Then try to lock it in the original thread. This would remove the thread park stuff.

Comment (garypen) [reply apollo-router/src/drop_watch.rs:22]: Because you can't `join` it. You have to `await`. This is specifically for code which is not async and which we would like to know it completed.

File: apollo-router/src/drop_watch.rs
Comment (garypen) [reply apollo-router/src/drop_watch.rs:24]: Because you can't `join` it. You have to `await`. This is specifically for code which is not async and which we would like to know it completed.

File: apollo-router/src/drop_watch.rs
Comment (garypen) [reply apollo-router/src/drop_watch.rs:15]: That's what I had originally, but the ManuallyDrop approach is cleaner. Like I said yesterday, `take` is only marked as `unsafe` because it's unsafe to do unless you know what you are doing. It's the best way to do it from a `Drop`, so that's what the comment and link to the documentation are recording in the `Drop` implementation.

**Review by BrynCooke — DISMISSED** [review]
Approved, but we should have a couple more reviewers before merging.

File: apollo-router/src/drop_watch.rs
Comment (BrynCooke) [inline apollo-router/src/drop_watch.rs:68]: Are we still not effectively blocking a tokio thread here?

Comment (garypen) [reply apollo-router/src/drop_watch.rs:68]: At line 69 we have to wait for the join() to occur, but we are only waiting for our own thread to terminate. That's safe as long as our thread terminates. If our thread never terminated (infinite loop or other logic error), then we would have that problem anyway.
Summary: We want to join our threads to make sure that they execute correctly. That's part of the purpose of this struct and this is how we enforce it.

Another way to think about what this code is doing is to compare it to `spawn_blocking()`. Spawn blocking is waiting for an `async` task to complete. If we `await` the spawned task, then we are blocked until the spawned task completes. This code is doing exactly the same thing, expect it's:
 - for synchronous tasks, not async tasks (e.g. dropping file watchers, shutting down tracing)
 - synchronous itself, so we can join where we can't `await`.

Running it from an async context would only have the same impact as running `spawn_blocking()` in an async context and the awaiting that task.

File: apollo-router/src/plugins/rhai/mod.rs
Comment (Geal) [inline apollo-router/src/plugins/rhai/mod.rs:150]: instead of a thread, that one could be moved into a spawn_blocking

File: apollo-router/src/drop_watch.rs
Comment (Geal) [inline apollo-router/src/drop_watch.rs:69]: I am not sure this is the behaviour we want. We need the various threads to be joined before stopping the process, but here that would be at the cost of an executor thread blocked waiting for the `DropWatch` to execute its stack. Which means that every time we reload schema or config, dropping the old tracer provider and meter provider would affect the async runtime

Comment (garypen) [reply apollo-router/src/drop_watch.rs:69]: It's hard to say which is worse: spawning unbound tasks/threads (complicated debugging) or possible synchronisation delays (are these going to be long in duration?). I think the former, but I can see there's a case for the latter and we should benchmark out the impact.

If we are trying to address the former, then the runtime would be affected however we do it. The constraint is that we wish to "join" (either `join` for sync context or `await` for async context) a spawned `thread`/`task`. Right now we don't do that and it is a source of debugging confusion and (potentially) a source of future bugs.

The main problem is trying to execute `sync` code from a `sync` fn which "lives within" an `async` call stack. We can't change the behaviour of `Drop`, so we need a way to be able to address the problem. This is the least intrusive mechanism I can concoct.

You could try to concoct something which enters the async runtime during `Drop`, but that might be executing after tokio has finished and you still won't have a way to `await` the spawned task. (You could work around the "runtime has finished" problem using `try_current()` but you still wouldn't have synchronisation point if there was a runtime.

Comment (Geal) [reply apollo-router/src/drop_watch.rs:69]: the main goal of the code before was to execute that synchronous task outside of the async executor. So joining on that thread in the async executor is a regression

Comment (garypen) [reply apollo-router/src/drop_watch.rs:69]: My point still stands. The existing code (telemetry) just spawned a thread and didn't join it in. The new code does the same thing but joins. We could take out the join, but then we'd be back where we were before but with a slightly tidier utility class.

Comment (Geal) [reply apollo-router/src/drop_watch.rs:69]: To rephrase my concerns here:
- dropping tracer providers in the async code was causing issues because it is a potentially long synchronous task, so it was moved to a thread
- I am not attached to the thread solution, and moving that pattern in a specific structure that encodes a correct behaviour sounds fine
- reintroducing the synchronous task in the async runtime is not a correct behaviour
- I agree that we need a way to track the threads we start, and we have to join on them at some point
- when to join on them is unclear to me at this point. Do you want to make sure that all threads have stopped before closing the process? Or is it during before switching to a new conf during a reload?
- as a side point, it is unclear to me how threads cause issues in tracking memory leaks with heaptrack, even with frequent reloads, but we can look into that elsewhere

**Comment by garypen** [comment]
Following further discussion with Geoffroy and Bryn, let's ignore this PR and consider an alternative approach that uses channels and:
 - drops the tracer 
 - reloads metrics
at the other end of the channel.

**Pull request closed without merging**
